src/main/cpp/src/config/pb_config.cc

```diff
@@ -0,0 +1,230 @@
+//Enable asserts
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <zlib.h>
+#include "pb_config.h"
+#include "tiledb_utils.h"
+#include "json_config.h"
+
+#define VERIFY_OR_THROW(X) if(!(X)) throw GenomicsDBConfigException(#X);
+
+void PBConfigBase::read_from_PB(const genomicsdb_pb::ExportConfiguration* export_config, const int rank) {
+  VERIFY_OR_THROW(export_config && export_config->IsInitialized());
+  read_and_initialize_vid_and_callset_mapping_if_available(export_config);
+  VERIFY_OR_THROW(m_vid_mapper.is_initialized() && m_vid_mapper.is_callset_mapping_initialized());
+  m_workspaces.clear();
+  m_workspaces.emplace_back(export_config->workspace());
+  m_single_workspace_path = true;
+  m_array_names.clear();
+  if(!export_config->has_array_name())
+    throw GenomicsDBConfigException("PB export config object must have array_name set");
+  m_array_names.emplace_back(export_config->array_name());
+  m_single_array_name = true;
+  if(export_config->has_scan_full() && export_config->scan_full())
+    scan_whole_array();
+  m_attributes.resize(export_config->attributes_size());
+  for(auto i=0;i<export_config->attributes_size();++i)
+    m_attributes[i] = std::move(std::string(export_config->attributes(i)));
+  if(export_config->has_query_filter())
+    m_query_filter = export_config->query_filter();
+  if(export_config->has_segment_size())
+    m_segment_size = export_config->segment_size();
+  if(export_config->has_vcf_header_filename())
+    m_vcf_header_filename = export_config->vcf_header_filename();
+  m_vcf_output_filename = export_config->has_vcf_output_filename()
+    ? export_config->vcf_output_filename() : "-"; //else stdout
+  m_vcf_output_format = export_config->has_vcf_output_format()
+    ? export_config->vcf_output_format() : "";
+  if(!m_scan_whole_array
+      && export_config->query_column_ranges_size() == 0
+      && export_config->query_contig_intervals_size() == 0
+      && export_config->query_row_ranges_size() == 0
+      && export_config->query_sample_names_size() == 0)
+    throw GenomicsDBConfigException("Must have one of \"scan_full\" or \"query_column_ranges\" or \
+	\"query_contig_intervals\" or \"query_row_ranges\" \
+	or \"query_sample_names\" in the Protobuf query config object");
+  if(!m_scan_whole_array) {
+    if(export_config->query_column_ranges_size() > 0 && export_config->query_contig_intervals_size() > 0)
+      throw GenomicsDBConfigException("Protobuf query object cannot have both query_column_ranges and query_contig_intervals at the same time");
+    if(export_config->query_column_ranges_size() > 0) {
+      if(export_config->query_column_ranges_size() == 1)
+	m_single_query_column_ranges_vector = true;
+      m_column_ranges.resize(export_config->query_column_ranges_size());
+      for(auto i=0;i<export_config->query_column_ranges_size();++i) {
+	const auto& inner_list_wrapper = export_config->query_column_ranges(i);
+	m_column_ranges[i].resize(inner_list_wrapper.column_or_interval_list_size());
```

Comment:
Maybe inline with the earlier question about GenomicsDBConfigBase vs VariantQueryConfig...if we did use VariantQueryConfig as the parent we could directly initialize m_query_column_intervals here instead of m_column_ranges, right? And similarly for m_query_rows and m_row_ranges below....

Reply:
I use the GenomicsDBConfigBase class as a 'bag' for the information provided by JSON/Protobuf (both import and query). The VariantQueryConfig members are somewhat curated - example, if the array is partitioned, then the m_query_column_intervals are cut at the partition bounds.
But, yes, you could rewrite the structures to avoid the duplication.